okio-fakefilesystem/src/commonMain/kotlin/okio/fakefilesystem/FakeFileSystem.kt

```diff
@@ -437,17 +438,51 @@ class FakeFileSystem(
   /** Reads data from [buffer], removing itself from [openPathsMutable] when closed. */
   private inner class FakeFileSource(
     private val openFile: OpenFile,
-    private val buffer: Buffer
-  ) : Source {
+    private val file: File
+  ) : Source, Cursor {
+    private val buffer = Buffer()
+    private var size = 0L
     private var closed = false
 
+    init {
+      seek(0L)
+    }
+
     override fun read(sink: Buffer, byteCount: Long): Long {
       check(!closed) { "closed" }
       return buffer.read(sink, byteCount)
     }
 
     override fun timeout() = Timeout.NONE
 
+    override fun cursor(): Cursor = this
+
+    override fun position(): Long {
+      check(!closed) { "closed" }
+      return size - buffer.size
+    }
+
+    override fun size(): Long {
+      check(!closed) { "closed" }
+      return size
+    }
+
+    override fun seek(position: Long) {
+      check(!closed) { "closed" }
```

Comment:
The thing I love here about FakeFileSystem is that the default behaviour is the most strict (allowReadsWhileWriting) which will flush out issues in testing, while being more lenient in production.  

Reviewing this code, I'm less certain whether I should worry about thread concurrency.

Should by default a FakeFileSource be locked to a single thread by default, simplifying reviewing? Potentially throwing if threads swap, or if we detect concurrency?

Reply:
Unfortunately that'll break producer consumer models where one thread picks files to read and another thread reads em.

But we could add an opt-in for callers that know they don't do that?

strict concurrency()

Reply:
My difficulty of the review is how to review whether this code is concurrency safe. Maybe on first read it makes the thread sticky is strictConcurrency is set?  Also in that case which thread creates the source instances?

Reply:
ZipFileSystem is intended to be immutable once created. Individual sources are not safe for concurrent use.